File: apps/worker/src/workflows/dynamic-dag.ts
import {
  proxyActivities, defineSignal, defineQuery, setHandler,
  condition, continueAsNew, workflowInfo,
} from '@temporalio/workflow';
import type {
  DynamicWorkflowInput, PipelineNode, PipelineEdge, NodeResult, DataRef, ExecutionStatus,
} from '@dataflow/shared';
import type * as activities from '../activities';

const act = proxyActivities<typeof activities>({
  startToCloseTimeout: '10 minutes',
  heartbeatTimeout: '60 seconds',
  retry: {
    initialInterval: '2s', backoffCoefficient: 2, maximumInterval: '5m',
    maximumAttempts: 5,
    nonRetryableErrorTypes: ['UnknownActivityError', 'SchemaValidationError'],
  },
});

export const pauseSignal  = defineSignal('pause');
export const resumeSignal = defineSignal('resume');
export const cancelSignal = defineSignal('cancel');
export const statusQuery  = defineQuery<ExecutionStatus>('status');

// One workflow interprets ALL user pipelines. The definition arrives frozen
// as input — replay-deterministic by construction.
export async function DynamicDAGWorkflow(input: DynamicWorkflowInput): Promise<ExecutionStatus> {
  const { definition, tenantId, executionId, trigger } = input;
  const results = new Map<string, NodeResult>();
  let paused = false, cancelled = false;

  setHandler(pauseSignal,  () => { paused = true; });
  setHandler(resumeSignal, () => { paused = false; });
  setHandler(cancelSignal, () => { cancelled = true; });
  setHandler(statusQuery, () => ({
    executionId,
    phase: cancelled ? 'cancelled' : paused ? 'paused' : 'running',
    nodeResults: Object.fromEntries(results),
    startedAt: new Date(workflowInfo().startTime).toISOString(),
  }));

  // ── Build execution plan: Kahn topological sort into parallel levels ──
  const { levels, incoming, outgoing } = plan(definition.nodes, definition.edges);
  const maxParallel = definition.concurrency?.maxParallelNodes ?? 5;

  for (const level of levels) {
    if (cancelled) break;
    await condition(() => !paused || cancelled);
    if (cancelled) break;

    for (let i = 0; i < level.length; i += maxParallel) {
      const batch = level.slice(i, i + maxParallel);
      await Promise.all(batch.map(node =>
        runNode(node, { definition, tenantId, executionId, trigger,
                        results, incoming, outgoing })));
    }
  }

  const failed = [...results.values()].some(r => r.status === 'failed');
  const phase = cancelled ? 'cancelled' : failed ? 'failed' : 'completed';
  await act.markExecution({ executionId, phase });
  return {
    executionId, phase: phase as any,
    nodeResults: Object.fromEntries(results),
    startedAt: new Date(workflowInfo().startTime).toISOString(),
    completedAt: new Date().toISOString(),
  };
}

interface RunCtx {
  definition: DynamicWorkflowInput['definition'];
  tenantId: string; executionId: string;
  trigger: DynamicWorkflowInput['trigger'];
  results: Map<string, NodeResult>;
  incoming: Map<string, PipelineEdge[]>;
  outgoing: Map<string, PipelineEdge[]>;
}

async function runNode(node: PipelineNode, ctx: RunCtx): Promise<void> {
  const { results, incoming } = ctx;
  const inEdges = incoming.get(node.id) ?? [];

  // upstream skipped/failed → skip
  const upstream = inEdges.map(e => results.get(e.source));
  if (inEdges.length && upstream.every(u => u?.status !== 'success')) {
    results.set(node.id, { nodeId: node.id, status: 'skipped', meta: { durationMs: 0 } });
    return;
  }

  // conditional edges
  for (const e of inEdges.filter(e => e.condition)) {
    const ok = await act.evalEdgeCondition({
      condition: e.condition!, inputRef: results.get(e.source)?.outputRef });
    if (!ok) {
      results.set(node.id, { nodeId: node.id, status: 'skipped', meta: { durationMs: 0 } });
      return;
    }
  }

  try {
    if (node.type === 'source') {
      results.set(node.id, await runSource(node, ctx));
    } else if (node.type === 'merge') {
      const refs = inEdges.map(e => results.get(e.source)?.outputRef).filter(Boolean) as DataRef[];
      results.set(node.id, await act.mergeRefs({
        inputRefs: refs, strategy: node.mergeStrategy ?? 'concat',
        joinKey: node.joinKey, tenantId: ctx.tenantId,
        executionId: ctx.executionId, nodeId: node.id }));
    } else { // transform | sink | fork(passthrough)
      const inputRef = node.type === 'fork'
        ? results.get(inEdges[0]?.source)?.outputRef
        : results.get(inEdges[0]?.source)?.outputRef
          ?? ctx.trigger.payloadRef;             // webhook payload feeds first transform
      if (node.type === 'fork') {
        results.set(node.id, { nodeId: node.id, status: 'success',
          outputRef: inputRef, meta: { durationMs: 0 } });
      } else {
        results.set(node.id, await act.dispatchNode({
          activityType: node.activityType, config: node.config,
          inputRef, tenantId: ctx.tenantId,
          executionId: ctx.executionId, nodeId: node.id }));
      }
    }
  } catch (err: any) {
    results.set(node.id, { nodeId: node.id, status: 'failed',
      meta: { durationMs: 0 }, error: err?.message ?? String(err) });
  }
}

// Sources page in a loop. Historical backfill may take thousands of pages —
// continueAsNew bounds Temporal history. Cursor state lives in Postgres,
// so the fresh workflow run picks up exactly where the old one stopped.
async function runSource(node: PipelineNode, ctx: RunCtx): Promise<NodeResult> {
  const connectionId = `${ctx.definition.id}:${node.id}`;
  let pages = 0, total = 0;
  let lastRef: DataRef | undefined;

  while (true) {
    const page = await act.fetchSourcePage({
      activityType: node.activityType, config: node.config,
      ingestion: node.ingestion, tenantId: ctx.tenantId,
      connectionId, executionId: ctx.executionId, nodeId: node.id,
    });
    lastRef = page.outputRef; total += page.recordCount; pages++;

    if (!page.hasMore) break;
    if (pages >= 50) {
      // long backfill: reset history, same input → cursor in PG resumes us
      await continueAsNew<typeof DynamicDAGWorkflow>({
        definition: ctx.definition, tenantId: ctx.tenantId,
        executionId: ctx.executionId, trigger: ctx.trigger });
    }
  }
  return { nodeId: node.id, status: 'success', outputRef: lastRef,
           meta: { durationMs: 0, recordCount: total } };
}

function plan(nodes: PipelineNode[], edges: PipelineEdge[]) {
  const inDeg = new Map<string, number>();
  const incoming = new Map<string, PipelineEdge[]>();
  const outgoing = new Map<string, PipelineEdge[]>();
  const byId = new Map(nodes.map(n => [n.id, n]));
  nodes.forEach(n => { inDeg.set(n.id, 0); incoming.set(n.id, []); outgoing.set(n.id, []); });
  edges.forEach(e => {
    outgoing.get(e.source)!.push(e);
    incoming.get(e.target)!.push(e);
    inDeg.set(e.target, (inDeg.get(e.target) ?? 0) + 1);
  });
  const levels: PipelineNode[][] = [];
  let q = nodes.filter(n => inDeg.get(n.id) === 0);
  let visited = 0;
  while (q.length) {
    levels.push(q); visited += q.length;
    const next: PipelineNode[] = [];
    for (const n of q) for (const e of outgoing.get(n.id)!) {
      const d = inDeg.get(e.target)! - 1;
      inDeg.set(e.target, d);
      if (d === 0) next.push(byId.get(e.target)!);
    }
    q = next;
  }
  if (visited !== nodes.length) throw new Error('Pipeline contains a cycle');
  return { levels, incoming, outgoing };
}

File: apps/workflow-go/internal/workflows/dynamic_dag.go
package workflows

import (
	"fmt"
	"time"

	"github.com/dataflow-poc/workflow-go/internal/model"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

type workflowState struct {
	Paused    bool
	Cancelled bool
	Results   map[string]model.NodeResult
}

type executionPlan struct {
	Levels   [][]model.Node
	Incoming map[string][]model.Edge
}

func DynamicDAGWorkflow(ctx workflow.Context, input model.WorkflowInput) (model.ExecutionStatus, error) {
	info := workflow.GetInfo(ctx)
	if input.ExecutionID == "" {
		input.ExecutionID = "exec-" + info.WorkflowExecution.RunID
	}
	if input.Environment == "" {
		input.Environment = model.Environment(info.Namespace)
	}
	if input.Trigger.FiredAt == "" {
		input.Trigger.FiredAt = workflow.Now(ctx).UTC().Format(time.RFC3339Nano)
	}

	activityOptions := workflow.ActivityOptions{
		TaskQueue:           "dynamic-activities-" + string(input.Environment),
		StartToCloseTimeout: 10 * time.Minute,
		HeartbeatTimeout:    time.Minute,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:        2 * time.Second,
			BackoffCoefficient:     2,
			MaximumInterval:        5 * time.Minute,
			MaximumAttempts:        5,
			NonRetryableErrorTypes: []string{"UnknownActivityError", "SchemaValidationError", "QuotaExceededError"},
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	if input.Trigger.Type == "cron" && !input.ExecutionPrepared {
		if input.PipelineRowID == "" {
			return model.ExecutionStatus{}, fmt.Errorf("scheduled workflow is missing pipelineRowId")
		}
		err := workflow.ExecuteActivity(ctx, "prepareScheduledExecution", map[string]interface{}{
			"executionId":   input.ExecutionID,
			"pipelineRowId": input.PipelineRowID,
			"tenantId":      input.TenantID,
			"environment":   input.Environment,
			"workflowId":    info.WorkflowExecution.ID,
			"runId":         info.WorkflowExecution.RunID,
		}).Get(ctx, nil)
		if err != nil {
			return model.ExecutionStatus{}, err
		}
		input.ExecutionPrepared = true
	}

	plan, err := buildPlan(input.Definition.Nodes, input.Definition.Edges)
	if err != nil {
		return model.ExecutionStatus{}, err
	}
	state := &workflowState{Results: map[string]model.NodeResult{}}
	startedAt := workflow.Now(ctx).UTC().Format(time.RFC3339Nano)

	if err := workflow.SetQueryHandler(ctx, "status", func() (model.ExecutionStatus, error) {
		phase := "running"
		if state.Cancelled {
			phase = "cancelled"
		} else if state.Paused {
			phase = "paused"
		}
		return model.ExecutionStatus{
			ExecutionID: input.ExecutionID,
			Phase:       phase,
			NodeResults: state.Results,
			StartedAt:   startedAt,
		}, nil
	}); err != nil {
		return model.ExecutionStatus{}, err
	}

	pauseCh := workflow.GetSignalChannel(ctx, "pause")
	resumeCh := workflow.GetSignalChannel(ctx, "resume")
	cancelCh := workflow.GetSignalChannel(ctx, "cancel")

	maxParallel := 5
	if input.Definition.Concurrency != nil && input.Definition.Concurrency.MaxParallelNodes > 0 {
		maxParallel = input.Definition.Concurrency.MaxParallelNodes
	}

	for _, level := range plan.Levels {
		drainSignals(pauseCh, resumeCh, cancelCh, state)
		for state.Paused && !state.Cancelled {
			selector := workflow.NewSelector(ctx)
			selector.AddReceive(resumeCh, func(c workflow.ReceiveChannel, _ bool) {
				c.Receive(ctx, nil)
				state.Paused = false
			})
			selector.AddReceive(cancelCh, func(c workflow.ReceiveChannel, _ bool) {
				c.Receive(ctx, nil)
				state.Cancelled = true
			})
			selector.Select(ctx)
		}
		if state.Cancelled {
			break
		}

		for start := 0; start < len(level); start += maxParallel {
			end := start + maxParallel
			if end > len(level) {
				end = len(level)
			}
			futures := make([]workflow.Future, 0, end-start)
			for _, node := range level[start:end] {
				node := node
				future, settable := workflow.NewFuture(ctx)
				workflow.Go(ctx, func(ctx workflow.Context) {
					result, runErr := runNode(ctx, input, node, plan.Incoming, state.Results)
					if runErr != nil {
						settable.SetError(runErr)
						return
					}
					settable.Set(result, nil)
				})
				futures = append(futures, future)
			}
			for index, future := range futures {
				var result model.NodeResult
				if err := future.Get(ctx, &result); err != nil {
					return model.ExecutionStatus{}, err
				}
				state.Results[level[start+index].ID] = result
			}
		}
	}

	phase := "completed"
	if state.Cancelled {
		phase = "cancelled"
	} else {
		for _, result := range state.Results {
			if result.Status == "failed" {
				phase = "failed"
				break
			}
		}
	}
	if phase == "completed" {
		cursors := make([]map[string]interface{}, 0)
		dedupeCheckpoints := make([]map[string]interface{}, 0)
		for _, result := range state.Results {
			checkpoint, ok := result.Meta["checkpoint"].(map[string]interface{})
			connectionID, hasID := result.Meta["connectionId"].(string)
			if ok && hasID {
				cursors = append(cursors, map[string]interface{}{"connectionId": connectionID, "checkpoint": checkpoint})
			}
			if checkpoint, ok := result.Meta["dedupeCheckpoint"].(map[string]interface{}); ok {
				dedupeCheckpoints = append(dedupeCheckpoints, checkpoint)
			}
		}
		if len(cursors) > 0 {
			if err := workflow.ExecuteActivity(ctx, "commitSourceCursors", map[string]interface{}{
				"tenantId": input.TenantID, "cursors": cursors,
			}).Get(ctx, nil); err != nil {
				return model.ExecutionStatus{}, err
			}
		}
		if len(dedupeCheckpoints) > 0 {
			if err := workflow.ExecuteActivity(ctx, "commitDedupeKeys", map[string]interface{}{
				"tenantId": input.TenantID, "checkpoints": dedupeCheckpoints,
			}).Get(ctx, nil); err != nil {
				return model.ExecutionStatus{}, err
			}
		}
	}
	if err := workflow.ExecuteActivity(ctx, "markExecution", map[string]interface{}{
		"executionId": input.ExecutionID,
		"phase":       phase,
	}).Get(ctx, nil); err != nil {
		return model.ExecutionStatus{}, err
	}
	return model.ExecutionStatus{
		ExecutionID: input.ExecutionID,
		Phase:       phase,
		NodeResults: state.Results,
		StartedAt:   startedAt,
		CompletedAt: workflow.Now(ctx).UTC().Format(time.RFC3339Nano),
	}, nil
}

func runNode(
	ctx workflow.Context,
	input model.WorkflowInput,
	node model.Node,
	incoming map[string][]model.Edge,
	results map[string]model.NodeResult,
) (model.NodeResult, error) {
	inEdges := incoming[node.ID]
	successful := false
	for _, edge := range inEdges {
		if upstream, ok := results[edge.Source]; ok && upstream.Status == "success" {
			successful = true
		}
	}
	if len(inEdges) > 0 && !successful {
		return skipped(node.ID), nil
	}

	for _, edge := range inEdges {
		if edge.Condition == "" {
			continue
		}
		upstream := results[edge.Source]
		var allowed bool
		err := workflow.ExecuteActivity(ctx, "evalEdgeCondition", map[string]interface{}{
			"condition":    edge.Condition,
			"inputRef":     upstream.OutputRef,
			"encryptedDek": input.EncryptedDEK,
		}).Get(ctx, &allowed)
		if err != nil {
			return failed(node.ID, err), nil
		}
		if !allowed {
			return skipped(node.ID), nil
		}
	}

	if node.Type == "source" {
		return runSource(ctx, input, node)
	}

	refs := make([]*model.DataRef, 0, len(inEdges))
	for _, edge := range inEdges {
		if ref := results[edge.Source].OutputRef; ref != nil {
			refs = append(refs, ref)
		}
	}
	if node.Type == "merge" {
		var result model.NodeResult
		err := workflow.ExecuteActivity(ctx, "mergeRefs", map[string]interface{}{
			"inputRefs":    refs,
			"strategy":     defaultString(node.MergeStrategy, "concat"),
			"joinKey":      node.JoinKey,
			"tenantId":     input.TenantID,
			"executionId":  input.ExecutionID,
			"nodeId":       node.ID,
			"encryptedDek": input.EncryptedDEK,
		}).Get(ctx, &result)
		if err != nil {
			return failed(node.ID, err), nil
		}
		return result, nil
	}

	var inputRef *model.DataRef
	if len(refs) > 0 {
		inputRef = refs[0]
	} else {
		inputRef = input.Trigger.PayloadRef
	}
	if node.Type == "fork" {
		return model.NodeResult{
			NodeID: node.ID, Status: "success", OutputRef: inputRef,
			Meta: map[string]interface{}{"durationMs": 0},
		}, nil
	}

	var result model.NodeResult
	err := workflow.ExecuteActivity(ctx, "dispatchNode", map[string]interface{}{
		"activityType": node.ActivityType,
		"config":       node.Config,
		"inputRef":     inputRef,
		"tenantId":     input.TenantID,
		"executionId":  input.ExecutionID,
		"nodeId":       node.ID,
		"encryptedDek": input.EncryptedDEK,
	}).Get(ctx, &result)
	if err != nil {
		return failed(node.ID, err), nil
	}
	return result, nil
}

func runSource(ctx workflow.Context, input model.WorkflowInput, node model.Node) (model.NodeResult, error) {
	total := 0
	refs := make([]*model.DataRef, 0)
	var checkpoint map[string]interface{}
	connectionID := sourceConnectionID(input.Definition.ID, node)
	for pageNumber := 1; ; pageNumber++ {
		var page struct {
			OutputRef   *model.DataRef         `json:"outputRef"`
			HasMore     bool                   `json:"hasMore"`
			RecordCount int                    `json:"recordCount"`
			Checkpoint  map[string]interface{} `json:"checkpoint,omitempty"`
		}
		err := workflow.ExecuteActivity(ctx, "fetchSourcePage", map[string]interface{}{
			"activityType": node.ActivityType,
			"config":       node.Config,
			"ingestion":    node.Ingestion,
			"tenantId":     input.TenantID,
			"connectionId": connectionID,
			"cursor":       checkpoint,
			"executionId":  input.ExecutionID,
			"nodeId":       node.ID,
			"encryptedDek": input.EncryptedDEK,
		}).Get(ctx, &page)
		if err != nil {
			return failed(node.ID, err), nil
		}
		if page.OutputRef != nil {
			refs = append(refs, page.OutputRef)
		}
		if page.Checkpoint != nil {
			checkpoint = page.Checkpoint
		}
		total += page.RecordCount
		// ponytail: bound history/payload size; a later execution resumes from the saved cursor.
		if !page.HasMore {
			break
		}
		if pageNumber >= 50 {
			return failed(node.ID, fmt.Errorf("source partition exceeded 50 pages; reduce backfill partition size or increase pageSize")), nil
		}
	}
	var outputRef *model.DataRef
	if len(refs) == 1 {
		outputRef = refs[0]
	} else if len(refs) > 1 {
		var merged model.NodeResult
		if err := workflow.ExecuteActivity(ctx, "mergeRefs", map[string]interface{}{
			"inputRefs": refs, "strategy": "concat", "tenantId": input.TenantID,
			"executionId": input.ExecutionID, "nodeId": node.ID,
			"encryptedDek": input.EncryptedDEK,
		}).Get(ctx, &merged); err != nil {
			return failed(node.ID, err), nil
		}
		outputRef = merged.OutputRef
	}
	meta := map[string]interface{}{"durationMs": 0, "recordCount": total}
	if checkpoint != nil {
		meta["checkpoint"] = checkpoint
		meta["connectionId"] = connectionID
	}
	return model.NodeResult{
		NodeID: node.ID, Status: "success", OutputRef: outputRef, Meta: meta,
	}, nil
}

func sourceConnectionID(pipelineID string, node model.Node) string {
	id := pipelineID + ":" + node.ID
	if node.Config["syncMode"] == "cdc" {
		id += ":cdc"
	}
	if node.Ingestion != nil && node.Ingestion.StateKey != "" {
		id += ":" + node.Ingestion.StateKey
	}
	return id
}

func buildPlan(nodes []model.Node, edges []model.Edge) (executionPlan, error) {
	inDegree := map[string]int{}
	incoming := map[string][]model.Edge{}
	outgoing := map[string][]model.Edge{}
	byID := map[string]model.Node{}
	for _, node := range nodes {
		inDegree[node.ID] = 0
		byID[node.ID] = node
	}
	for _, edge := range edges {
		if _, ok := byID[edge.Source]; !ok {
			return executionPlan{}, fmt.Errorf("edge references unknown source %q", edge.Source)
		}
		if _, ok := byID[edge.Target]; !ok {
			return executionPlan{}, fmt.Errorf("edge references unknown target %q", edge.Target)
		}
		inDegree[edge.Target]++
		incoming[edge.Target] = append(incoming[edge.Target], edge)
		outgoing[edge.Source] = append(outgoing[edge.Source], edge)
	}
	queue := make([]model.Node, 0)
	for _, node := range nodes {
		if inDegree[node.ID] == 0 {
			queue = append(queue, node)
		}
	}
	levels := make([][]model.Node, 0)
	visited := 0
	for len(queue) > 0 {
		level := append([]model.Node{}, queue...)
		levels = append(levels, level)
		visited += len(level)
		next := make([]model.Node, 0)
		for _, node := range level {
			for _, edge := range outgoing[node.ID] {
				inDegree[edge.Target]--
				if inDegree[edge.Target] == 0 {
					next = append(next, byID[edge.Target])
				}
			}
		}
		queue = next
	}
	if visited != len(nodes) {
		return executionPlan{}, fmt.Errorf("pipeline contains a cycle")
	}
	return executionPlan{Levels: levels, Incoming: incoming}, nil
}

func drainSignals(
	pause workflow.ReceiveChannel,
	resume workflow.ReceiveChannel,
	cancel workflow.ReceiveChannel,
	state *workflowState,
) {
	for pause.ReceiveAsync(nil) {
		state.Paused = true
	}
	for resume.ReceiveAsync(nil) {
		state.Paused = false
	}
	for cancel.ReceiveAsync(nil) {
		state.Cancelled = true
	}
}

func skipped(nodeID string) model.NodeResult {
	return model.NodeResult{
		NodeID: nodeID, Status: "skipped",
		Meta: map[string]interface{}{"durationMs": 0},
	}
}

func failed(nodeID string, err error) model.NodeResult {
	return model.NodeResult{
		NodeID: nodeID, Status: "failed", Error: err.Error(),
		Meta: map[string]interface{}{"durationMs": 0},
	}
}

func defaultString(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
